### DISCUSSION TITLE
Which optimizer is better to use for training ASR models?

### DISCUSSION THREAD

**Discussion by ahkarami** [NVIDIA-NeMo/NeMo#3964] [Q&A] [answered]
@titu1994 
Thanks for your great repo. I have a question about optimizer for training ASR models. I have seen in the config yaml files of NeMo that some models (e.g., [Citrinet](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/citrinet/citrinet_512.yaml#L385)) trained via `novograd`, and some other (e.g., [Conformer_transducer_bpe](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/conformer/conformer_transducer_bpe.yaml#L193)) trained via `adamw`, and other (e.g., [Contextnet_rnnt](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/contextnet_rnnt/config_rnnt_bpe.yaml#L211)) trained via `adam`.
Precisely, my question is according to your experiences & experiments which optimizer generally better to use? 
I think, according to e.g., [this doc](https://nvidia.github.io/OpenSeq2Seq/html/optimizers.html#novograd), NovoGrad is generally better than Adam (takes less memory, and to be more numerically stable, compared to Adam). 
If this claim is true and you agree with me, then why have you not always used this optimizer (i.e., NovoGrad)? and for example for training Conformer model you have used AdamW?

**Comment by titu1994** [comment, marked as answer]
All optimizers work with all models. Some work much better though. 
From our experiments, we found that 
Citrinet / ContextNet do much better with Novograd + Cosine Annealing Scheduler as compared to Adam/w + Noam or Adam/w with Cosine Annealing Schedulers. 
On the other hand, we found that Conformer (CTC / RNNT) do much better with Adam/w with Noam as compared to Novograd with Cosine Annealing. 

The best combination can not only yield better final wer but also impact speed of convergence as well as stability of training and finetuning. As is all things in deep learning, the answer to your questions is : "it depends on many many interacting factors". 

Side note: Novograd requires less memory yes, but also normalized gradients, which generally leads to slower convergence. However the slower convergence usually yields very good WER for certain models (such as very deep CNN's - Jasper, QuartzNet, Citrinet, ContextNet family of models)
On the other hand, transformer-like models require adaptive learning rates controlled by second order moving averages for more stable training of dense self attention layers, and normalized gradients my slow down convergence too much, so Adam and noam scheduler do a better job for the Conformer family of models. 

There's no theoretical proof of the above statements, at least none that I know, but empirically these are the results we have obtained after hundreds of experiments per model architecture.

**Reply by ahkarami** [reply]
Thanks for your complete answer.
Best
